File: solutions/python/trees/kth-smallest-element-in-a-bst.py
# LeetCode 230 — Kth Smallest Element in a BST (Medium)
# Category: Trees · Approach: Inorder
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/kth-smallest-element-in-a-bst/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def kth_smallest(root: TreeNode | None, k: int) -> int:
    order: list[int] = []
    def inorder(node):
        if not node or len(order) >= k:
            return
        inorder(node.left)
        order.append(node.val)
        inorder(node.right)
    inorder(root)
    return order[k - 1]

File: solutions/typescript/trees/kth-smallest-element-in-a-bst.ts
// LeetCode 230 — Kth Smallest Element in a BST (Medium)
// Category: Trees · Approach: Inorder
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/kth-smallest-element-in-a-bst/

class TreeNode {
  val: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;
  constructor(val = 0) {
    this.val = val;
  }
}

function kthSmallest(root: TreeNode | null, k: number): number {
  const order: number[] = [];
  function inorder(node: TreeNode | null): void {
    if (!node || order.length >= k) return;
    inorder(node.left);
    order.push(node.val);
    inorder(node.right);
  }
  inorder(root);
  return order[k - 1];
}
